odl/space/cu_ntuples.py

```diff
@@ -847,36 +976,271 @@ def _norm_default(x):
     return x.data.norm()
 
 
+def _pnorm_default(x, p):
+    if p == float('inf'):
+        raise NotImplementedError('inf-norm not implemented.')
+    elif int(p) != p:
+        raise NotImplementedError('non-integer powers not implemented')
+    # TODO: optimized version in C++ code?
+    xp = abs(x)
+    xp **= p
+    return sum(xp)**(1/p)
+
+
+def _pnorm_diagweight(x, p, w):
+    if p == float('inf'):
+        raise NotImplementedError('inf-norm not implemented.')
+    elif int(p) != p:
+        raise NotImplementedError('non-integer powers not implemented')
+    # TODO: optimized version in C++ code?
+    xp = abs(x)
+    xp **= p
+    xp *= w
+    return sum(xp)**(1/p)
+
+
 def _inner_default(x1, x2):
     return x1.data.inner(x2.data)
 
 
-class _CudaFnWeighting(with_metaclass(ABCMeta, _FnWeightingBase)):
+class _CudaFnWeighting(_FnWeightingBase):
 
     """Abstract base class for `CudaFn` weighting."""
 
 
-class _CudaFnConstWeighting(_CudaFnWeighting):
+class CudaFnVectorWeighting(_CudaFnWeighting):
+
+    """Vector weighting for `CudaFn`.
+
+    For exponent 2.0, a new weighted inner product with vector :math:`w`
+    is defined as
+
+    :math:`<a, b>_w := b^H (w * a)`
+
+    with :math:`b^H` standing for transposed complex conjugate, and
+    `w * a` being element-wise multiplication.
+
+    For other exponents, only norm and dist are defined. In the case of
+    exponent `inf`, the new norm is equal to the unweighted one,
+
+    :math:`||a||_{w, \infty} := ||a||_\infty`,
+
+    otherwise it is
+
+    :math:`||a||_{w, p} := ||w^{1/p} * a||_p`.
+
+    The vector may only have positive entries, otherwise it does not
+    define an inner product or norm, respectively. This is not checked
+    during initialization.
+    """
+
+    def __init__(self, vector, exponent=2.0, dist_using_inner=False,
+                 copy_to_gpu=False):
```

Comment:
Having thought about this again, it's probably better to not allow this copy (and remove the option) but to require giving the initializer directly a `CudaFn.Vector`. Otherwise, the behavior of the `__eq__()` is weird: If a NumPy array is provided, and `copy_to_gpu` is `True`, two weightings initialized with the identical array will still not be equal.
The same is true a `CudaFn.Vector` is used in init with wrong data type, but that can be better justified (and documented).